config_ext_test.go

```diff
@@ -0,0 +1,350 @@
+nothing: ~
+fun:
+  - maserati
+  - porsche
+practical:
+  toyota: camry
+  honda: accord
+occupants:
+  honda:
+    driver: jane
+    backseat: [nate]
+`)
+	second := strings.NewReader(`
+fun:
+  - maserati
+  - lamborghini
+practical:
+  honda: civic
+  nissan: altima
+occupants:
+  honda:
+    passenger: arthur
+    backseat: [nora]
+`)
+
+	provider, err := NewYAMLProviderFromReader(first, second)
+	require.NoError(t, err, "couldn't create provider")
+	assert.Equal(t, `cached "yaml"`, provider.Get(Root).Source(), "wrong source")
+	assert.Equal(t, "<nil>", provider.Get("nothing").String(), "wrong string representation")
+
+	type TestCase struct {
+		Value     Value
+		HasValue  bool
+		Interface interface{}
+		Populate  interface{}
+	}
+
+	run := func(tt TestCase) {
+		assert.Equal(t, tt.HasValue, tt.Value.HasValue(), "unexpected result from HasValue")
+		assert.Equal(t, tt.Interface, tt.Value.Value(), "unexpected result from Value")
+		require.NoError(t, tt.Value.Populate(tt.Populate), "error populating")
+	}
+
+	t.Run("missing", func(t *testing.T) {
+		var s string
+
+		t.Run("top level", func(t *testing.T) {
+			run(TestCase{
+				Value:     provider.Get("not_a_key"),
+				HasValue:  false,
+				Interface: nil,
+				Populate:  &s,
+			})
+			assert.Equal(t, "", s, "unexpected result after Populate")
+		})
+
+		t.Run("subkey", func(t *testing.T) {
+			run(TestCase{
+				Value:     provider.Get("practical.cadillac"),
+				HasValue:  false,
+				Interface: nil,
+				Populate:  &s,
+			})
+			assert.Equal(t, "", s, "unexpected result after Populate")
+		})
+
+		t.Run("descend into slice", func(t *testing.T) {
+			run(TestCase{
+				Value:     provider.Get("fun.extra"),
+				HasValue:  false,
+				Interface: nil,
+				Populate:  &s,
+			})
+			assert.Equal(t, "", s, "unexpected result after Populate")
+		})
+	})
+
+	t.Run("nil", func(t *testing.T) {
+		var s string
+		run(TestCase{
+			Value:     provider.Get("nothing"),
+			HasValue:  true,
+			Interface: nil,
+			Populate:  &s,
+		})
+		assert.Equal(t, "", s, "unexpected result after Populate")
+	})
+
+	t.Run("scalar", func(t *testing.T) {
+		var s string
+		run(TestCase{
+			Value:     provider.Get("practical.toyota"),
+			HasValue:  true,
+			Interface: "camry",
+			Populate:  &s,
+		})
+		assert.Equal(t, "camry", s, "unexpected result after Populate")
+	})
+
+	t.Run("map", func(t *testing.T) {
+		var m map[string]string
+		run(TestCase{
+			Value:    provider.Get("practical"),
+			HasValue: true,
+			Interface: map[interface{}]interface{}{
+				"toyota": "camry",
+				"honda":  "civic",
+				"nissan": "altima",
+			},
+			Populate: &m,
+		})
+		assert.Equal(t, map[string]string{
+			"toyota": "camry",
+			"honda":  "civic",
+			"nissan": "altima",
+		}, m, "unexpected result after populate")
+	})
+
+	t.Run("slice", func(t *testing.T) {
+		var s []string
+		run(TestCase{
+			Value:     provider.Get("fun"),
+			HasValue:  true,
+			Interface: []interface{}{"maserati", "lamborghini"},
```

Comment:
for some reason, i thought that the custom merging logic merged lists, rather than replacing, but i guess it matches the yaml unmarshaler?

Reply:
Yep. There are a bunch of internal tickets where we discussed merging lists, but the actual implementation appears to match the yaml package pretty closely.